src/main/java/org/elasticsearch/action/admin/indices/seal/TransportSealIndicesAction.java

```diff
@@ -0,0 +1,168 @@
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.GroupShardsIterator;
+import org.elasticsearch.cluster.routing.ShardIterator;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.SyncedFlushService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.*;
+
+import java.io.IOException;
+import java.util.Set;
+
+/**
+ */
+public class TransportSealIndicesAction extends HandledTransportAction<SealIndicesRequest, SealIndicesResponse> {
+
+
+    final private SyncedFlushService syncedFlushService;
+    final private ClusterService clusterService;
+    private final TransportService transportService;
+
+    @Inject
+    public TransportSealIndicesAction(Settings settings, ThreadPool threadPool, TransportService transportService, ActionFilters actionFilters, SyncedFlushService syncedFlushService, ClusterService clusterService) {
+        super(settings, SealIndicesAction.NAME, threadPool, transportService, actionFilters, SealIndicesRequest.class);
+        this.syncedFlushService = syncedFlushService;
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        transportService.registerRequestHandler(SealIndicesAction.NAME, ShardSyncedFlushRequest.class, ThreadPool.Names.GENERIC, new ShardSyncedFlushHandler());
+    }
+
+
+    @Override
+    protected void doExecute(final SealIndicesRequest request, final ActionListener<SealIndicesResponse> listener) {
+        // we need to execute in a thread here because getSyncedFlushResults(request, state) blocks
+        ClusterState state = clusterService.state();
+        String[] concreteIndices = state.metaData().concreteIndices(request.indicesOptions(), request.indices());
+        GroupShardsIterator primaries = state.routingTable().activePrimaryShardsGrouped(concreteIndices, false);
+        final CountDown countDown = new CountDown(primaries.size());
+        DiscoveryNode localNode = state.getNodes().localNode();
+        final Set<SyncedFlushService.SyncedFlushResult> results = ConcurrentCollections.newConcurrentSet();
+        for (ShardIterator shard : primaries) {
+            final ShardId shardId = shard.shardId();
+            transportService.sendRequest(localNode, SealIndicesAction.NAME, new ShardSyncedFlushRequest(shardId),
```

Comment:
I really wonder why we need this transport handler here at all. The `SyncedFlushService` already does all the remote work for us if the primary is not local. All you need to do is to call `syncedFlushService.attemptSyncedFlush` for each shard in the `GroupShardsIterator primaries` and pass an action listener to it that returns once all shards have returned?